### PR TITLE
hkj-spring: the 422 leg — one response, every bad field by path (#627)

### LINKED ISSUE #627: [FEAT] @GenerateMapping × hkj-spring: the 422 leg — one response, every bad field by path
## The gap

The mapper's promise ([#600](https://github.com/higher-kinded-j/higher-kinded-j/issues/600)) includes "a fallible `parse` drops onto the `ValidationPath` railway and into `hkj-spring` — one 422, every bad field by path." Step-0 ([#605](https://github.com/higher-kinded-j/higher-kinded-j/pull/605)) and the remainder ([#621](https://github.com/higher-kinded-j/higher-kinded-j/pull/621)) deliver the located, accumulating `parse`; the **Spring boundary integration** that turns a `Validated<NonEmptyList<FieldError>, Domain>` into a single `422 Unprocessable Entity` with a field-keyed body is not yet wired.

## What "good" looks like (user perspective)

```java
@PostMapping("/users")
Object create(@RequestBody UserDto dto) {
  return UserMappingImpl.INSTANCE.parse(dto);   // Validated<NonEmptyList<FieldError>, User>
}
// Invalid -> HTTP 422, body: { "errors": [ { "path": "email",       "message": "not an email address" },
//                                          { "path": "address.zip", "message": "must be 5 digits" } ] }
// Valid   -> the handler proceeds with the parsed User
```

- **A return-value handler / advice** maps an `Invalid` parse to 422 with every `FieldError` rendered by `pathString()`, consistent with the existing `*PathReturnValueHandler`s and the `EitherOrBoth` warnings-in-header handler.
- **Jackson round-trips** the error body using the existing `hkj-spring` support ([#578](https://github.com/higher-kinded-j/higher-kinded-j/pull/578)).
- **Opt-in and additive** — controllers that don't return a mapper `parse` result are untouched.

## Things to explore / constraints

- **Where it hooks.** A dedicated handler for `Validated<NonEmptyList<FieldError>, T>` returns, or reuse of the `ValidationPath` handler; confirm status (422 vs 400) and the error-body schema (align with any existing error envelope).
- **Path rendering.** Use `FieldError.pathString()`; carry the dotted-key caveat from #621 (a key containing a dot is indistinguishable from nesting in the *rendered* string; the structured path stays exact).
- **Content negotiation & message i18n** are design points, likely out of the first slice.
- **Testing** against `hkj-spring`'s existing web-layer harness.

## Dependencies / scope

- **Depends on** the located `parse` (#605/#621), `hkj-spring` generic deserializers ([#578](https://github.com/higher-kinded-j/higher-kinded-j/pull/578)), and the existing Path return-value handlers.
- **In:** a Spring integration rendering a fallible mapper `parse` as one 422 with per-field, path-located errors.
- **Out:** message i18n; a bespoke problem+json schema beyond the field list (can follow).

### PR DESCRIPTION
**Pull request by MagnusSmith**
Closes #627.

A controller returning a fallible mapper `parse` — `Validated<NonEmptyList<FieldError>, Domain>`, directly or via `ValidationPath` — now renders as **one 422 Unprocessable Content response listing every bad field by path**. Design authority: the [design-options comment](https://github.com/higher-kinded-j/higher-kinded-j/issues/627#issuecomment-5037964494) on the issue.

```json
{ "valid": false,
  "errors": [
    { "path": "email",       "segments": ["email"],          "message": "not a valid email address" },
    { "path": "address.zip", "segments": ["address", "zip"], "message": "must be 5 digits" }
  ],
  "errorCount": 2 }
```

## How it works

`ValidationPathReturnValueHandler` shape-checks the materialised Invalid payload: a `NonEmptyList`, `Collection` or array made **entirely** of located `FieldError`s — or one bare `FieldError` — takes the new leg with `hkj.web.validation-field-error-status` (default 422); each error renders as `{path, segments, message}`, where `path` is the dot-joined display key and `segments` the exact structured location (a key containing a dot is indistinguishable from nesting in `path` — the #621 caveat — so `segments` stays lossless). Mixed, empty or non-`FieldError` payloads keep the pre-existing generic rendering and `validation-invalid-status` (default 400), byte-for-byte. No Jackson-module change: `FieldError`'s exact wire shape is untouched everywhere else, and the `{valid, errors, errorCount}` family envelope is kept.

## Behaviour change (deliberate, pre-1.0)

The leg is selected **by shape, not source**: `Path.fields()` assemblies, `@GenerateAssembly` results and hand-built `NonEmptyList<FieldError>`s take it too. A pre-existing endpoint returning that shape moves 400 → 422 on upgrade; `hkj.web.validation-field-error-status: 400` restores the old status. Documented in the book section and CONFIGURATION.md.

## Changes

- **Handler** (`hkj-spring:autoconfigure`): shape detection + rendering in `ValidationPathReturnValueHandler`; package-private `FieldErrorItem` record with `FieldError`-style canonical-constructor discipline; backward-compatible 2-arg constructor kept per the Either siblings' precedent.
- **Config**: `HkjProperties.Web.validationFieldErrorStatus` (default 422), threaded through `HkjWebMvcAutoConfiguration` like the sibling status knobs.
- **Showcase** (`hkj-spring:example`): `POST /api/users/parse` returns `UserMappingImpl.INSTANCE.parse(dto)` directly — full `UserMapping` spec over a **bean-shaped** `UserDto`, so an absent JSON field becomes a located `"must not be null"` error in the same 422 rather than an exception (bean wires null-guard every read; record wires cannot). The PATCH endpoint deliberately stays on the Either leg (it mixes not-found with validation), cross-noted in javadoc.
- **Tests**: full edge-case table in `ValidationPathReturnValueHandlerTest` (shape detection across payload types, mixed-payload fallback, unlabelled errors, dotted-key round-trip rebuilding the original `FieldError` from `segments`, status-knob selection, back-compat constructor); property-binding tests; MockMvc slice test driving the leg end to end.
- **Docs**: `CONFIGURATION.md` + a "The 422 leg" book section (cross-linked from the record-mapping chapter), including the Spring-`FieldError` disambiguation and unlabelled-error rendering; hkj-spring skill refreshed.
- **Hygiene**: repo-wide `HttpStatus.UNPROCESSABLE_ENTITY` → `UNPROCESSABLE_CONTENT` (deprecated in Spring Framework 7 per RFC 9110); prose standardised on "Unprocessable Content".

## Review

Five-lens expert-panel review (correctness, type safety, immutability, ergonomics, HKJ architecture) ran post-implementation; 5 majors + 5 minors confirmed and fixed in the three `review fixes` commits — including the showcase NPE on absent fields (found independently by two lenses, confirmed empirically) and the property rename to match the `{handler}-{condition}-status` grammar.

Full hkj-spring suite + spotless green. `segments` stays a flat string list until the queued sealed `PathSegment` model lands; the richer schema is explicitly deferred to that issue.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Located `FieldError` validation failures now render via a dedicated **422 Unprocessable Content** “422 leg” as one aggregated response.
  * Added `hkj.web.validation-field-error-status` to configure the **422 leg** status (defaults to 422), separated from the generic invalid validation status.
* **Documentation**
  * Updated Spring Boot integration/configuration and “Record Mapping” guidance; expanded the glossary with **Leg** terminology.
* **Bug Fixes**
  * Located field-error JSON consistently includes `path`, `segments`, and `message`.
* **Tests**
  * Added slice and integration tests covering the default **422** behavior and the new property interactions.

### CODE REVIEW

File: hkj-spring/example/src/test/java/org/higherkindedj/spring/example/controller/UserParseWebMvcSliceTest.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+                    {"id":"7","email":"not-an-email","firstName":"","lastName":"Hopper"}
+                    """))
+        .andExpect(status().isUnprocessableContent())
+        .andExpect(jsonPath("$.valid").value(false))
+        .andExpect(jsonPath("$.errorCount").value(2))
+        .andExpect(jsonPath("$.errors[*].path").value(containsInAnyOrder("email", "firstName")))
+        .andExpect(
+            jsonPath("$.errors[?(@.path=='email')].message").value("not a valid email address"))
+        .andExpect(jsonPath("$.errors[?(@.path=='email')].segments[0]").value("email"))
+        .andExpect(jsonPath("$.errors[?(@.path=='firstName')].message").value("must not be blank"));
```
[the remark this answers is not in the document]
Comment (MagnusSmith) [reply hkj-spring/example/src/test/java/org/higherkindedj/spring/example/controller/UserParseWebMvcSliceTest.java:78]: These assertions pass — `JsonPathExpectationsHelper.assertValue` supports indefinite paths by asserting containment when the expected value is not a list, and `UserParseWebMvcSliceTest` runs green in CI and locally (4/4). Each filter matches at most one error by construction, so containment and `contains(...)` are equivalent here.

File: hkj-spring/example/src/main/java/org/higherkindedj/spring/example/controller/UserMapping.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ * 422 leg (issue #627): a controller returns the generated {@code
+ * UserMappingImpl.INSTANCE.parse(dto)} directly, and the {@code ValidationPathReturnValueHandler}
+ * renders an invalid parse as one {@code hkj.web.field-error-status} response (default 422
+ * Unprocessable Entity) carrying every located {@code FieldError} by path.
+ *
+ * <p>{@code email} and {@code firstName} are validated through leaves, so a request with both bad
+ * accumulates both errors in a single response; {@code id} and {@code lastName} copy by identity.
+ * The wire side is a bean, so every reference read is null-guarded by the generated {@code parse}:
+ * an absent field becomes a located {@code FieldError} ({@code "must not be null"}) and a {@code
+ * null} never reaches a leaf.
```
[the remark this answers is not in the document]
Comment (MagnusSmith) [reply hkj-spring/example/src/main/java/org/higherkindedj/spring/example/controller/UserMapping.java:23]: Fixed in 65655473d — the class-level javadoc now names `hkj.web.validation-field-error-status` and the RFC 9110 label "Unprocessable Content", matching the rest of the branch. A repo-wide sweep confirms this was the last stale mention.

**Comment by MagnusSmith** [comment]
@coderabbitai review

File: hkj-book/src/spring/spring_boot_integration.md [hunk 3]
```diff
@@ -367,9 +368,41 @@ public ValidationPath<NonEmptyList<ValidationError>, User> validateAndCreate(Use
...
+  ],
+  "errorCount": 2
+}
+```
+
+`path` is the dot-joined display key (`FieldError.pathString()`); `segments` is the exact structured location. The distinction matters: a map key containing a dot is indistinguishable from nesting in the rendered `path` (`attributes."a.b".email` vs deeper nesting), so structured consumers should read `segments`, which stays exact. An unlabelled error renders `"path": ""` and `"segments": []` — treat it as object-level.
+
+The leg is selected purely by payload shape, so **any** all-`FieldError` payload takes it — `Path.fields()` assemblies, `@GenerateAssembly` results and hand-built `NonEmptyList<FieldError>`s included, not just a mapper `parse`. A pre-existing `Path.fields()` endpoint therefore moves from 400 to 422 on upgrade; set `hkj.web.validation-field-error-status: 400` to restore the old status.
+
+Mixed or non-`FieldError` payloads keep the generic rendering and `hkj.web.validation-invalid-status` (default 400) — nothing to enable, nothing else affected.
```
[the remark this answers is not in the document]
Comment (MagnusSmith) [reply hkj-book/src/spring/spring_boot_integration.md]: Good catch — the handler javadoc documented the empty-payload fallback but the reader-facing docs didn't. Added "empty" to both fallback sentences (book + CONFIGURATION.md) and pinned the behaviour with a new `emptyPayloadFallsBack` unit test (`Validated.invalid(List.of())` → 400, generic rendering) in 054c15021.
